core/src/avm2/optimize.rs

```diff
@@ -563,6 +559,8 @@ pub fn optimize<'gc>(
                         || value2.class == Some(types.number))
                 {
                     stack.push_class(types.number);
+                } else if value1.class == Some(types.string) || value2.class == Some(types.string) {
```

Comment:
I don't think it's always correct:
```
            var a: String = null;
            var b: String = null;
            var c = a + b;
            trace(c); // 0
            trace(getQualifiedClassName(c)); // int
```
In FP's verifier, this optimization is present, but also guarded by `notNull` abstract value metadata checks (which we don't model currently, we only model "guaranteed null".)

Reply:
We could add a trivial `guaranteed_not_null` that's `true` for `pushstring` opcodes, which would allow type propagation for `"asdf" + whatever`